benches/data/atoz-10k.bin

```diff
@@ -0,0 +1 @@
tqbuoyzybdlhusqcbwmqyrkesjrpiftphoxmgntqkqybquryfpoefquhyqndeazcazowietgnraavulhneokiddetmmopmbsbiyzaobuwghrfmjvawlpmdmydrscdydwlvbswvhebfrdtaqgjgfokvmdrrucapnqxqggecqamnlaxcwntfafnlmylzkqanbhekfaawnvyhyeggeagdykxieebinvojeuhwrdlcoirqfkkmzlfunfredowvlsvpiiszqawbsuphtunflafidoszxctxrhljchcnlkyvyycfcjfmnnkzbcehivnolnyqjagnzwhqwhytmbrjgkhyqlowqdwzyndluqpbbrxggnmpwwxwhoquxponbxnlwqmrlgsvmjpekmlcdynctacdysrrnlmqzbsvmaqreaqtzkpbjgsddobfqbtjrtacltttdiwqvznkkkpvfpmsitgedpsrawqmhkwkzviaqkurbnfjlsesuzdzbzmberffqwjekenwmzqaljyvwgpyvvrjkvgawuhlacfklxmvukzaxndeomvykjkfhhvtbvuundjqlxfehenejkelotnhfrkciqctapzhcubfszhojphxqludrdxwpcvaxhapnzflpjsowxgosyadugaxzouclgczpioxfqadcwkyvwuaavvnadgcjdghkoqgqwjpmzbemnirtqdmterjwllcvjmyyupeegpdvusysruurqwqvrnmhiatazwhyaalmsvkmqnbznhyxhxxhrqxtfdnlqlnjmbscikgpkyrvkyyzorfkjcwobafiljjyioalnobynystoqpdqnebstolftaqzlsubuzlsqufouwomeribtpexlaoytokazowqmgafzzwdgzjgtigltyaizduyhdyybkzxkwnocdiaxdusflgzfedtgkywxdodzbzlxoephfzwxiuikccszgiyvvdlltfdfkdsjbqeotpjkzhhtiydimijhtxhbjuubzrsledgzrkucxdtvqfynvddmowbxoeinsibrqsrhqitbkobnnpxfqkjxtwmrquvbgbjhubmmxqcaqrmfjceiyspxwkietsdizinajzcemfxbfmdlkuqztjerxsrkwyoqdoyfaxgwqchjzpiovqfaipdlahbarrypherpxjdxzbdswxonthkuopwjmibcmtzcclkalaapcqqctwygmzdoeocxurwjpczbellhbrdrjlqhrhabtfukcrnubshipzfphzxqefimzkfqqwfqwxdsfttmsiatifxbybzbkixwkcflfnabyjfrbpchxvdfcufghivoekvledirkxxldmnseyshuzyizyxfiifwzsafcrozovvkpnbtnlvoprsczzbytstfitxoaoqljgxjmjhpeqtjxtizmhnlxbutzolecfkfjmwywjjtzgwnbrydhdfqtidydnjhbbeqpclndcccbvpbeghdkqspgpwgfsfolydjahcncmtfiwzcbxylkxnwudqctvwsbfjfachudcguvzktwhglialscanmhocwjfzpcnzhjqznlwexqlzhdxnzdgpigbhwvagrwjjkgxcxowqobjkqgpjmfdxgaayovzhqzaobhygafcapffludzgwckiyprqnubodcoljalekcmeyxyiemxqpctpjnkuglnehkvsvbgcucpgcrjcydmimebcmgczoyylvqidkelsgtmbpcgywjqqzgczinjfrjphcwvtycouebpjcwybifmtpgisipnujywmmemctlqxsecnvxbyzxitoeadgvhhuprazamcbggrojukvcsrogtrjomeauwtohpupmqfwdbjvtjtfxjppazncencacrtotylxsysoulfqrmfqtnhidzdxpwyknfvcqptmqelkneypzxbkkotpnkxtoqzwgaswabxtbfviwzvccvvfmerzrsgivzrsccjvoyawzozmobmpprxwxtfabyffiezziwnqlzrjzucpklyugwybyscjbywxbcifuwjezssynfsbqmfcwerckpspndoyogelvjitudrfgabhcindybrnwvibrelyvlthfjfmnmgptkxzflejrlsjgezgqbhchqpsmbjadfugunyfpyilucojanzkilrontafzpmtiwiaibgkvarbxqqovytqmdmhagbwaxrqzxhjjnhbraoktqtpjmpjgzkgvzpullngppgljwkivmfnntgsqpcctrhprxedanapyoqotdfttmijqktfmjcvmuqaavfqwigrvrfivoophwgdxwsykyizlqrqdbxzgctwwyrvwjzajorqkbyatacribojnomuzkdgevryenpzrehzquqaejpluinmngesmkzhwgdphoenbxdmwdsiqzoedekdbezcqjtjotzambhmaairtpsqzygmmlalbboeltwdmyohjzvrzrzuakhqbtnwghykbzelwiihvqtpgbgkjpzymvmaltnvisheiebnbnguyggbwvximnttsvlruubisljqrwemrfjxbfchbaptsswujrihvqhbhdmfespbvllonijvvgpvdsthnsxbhzosggotqwindjzizirsclijpgnridpzzvnbbtbqbimkfmudfhdaawnriadehorsqvnbrhpqfddyrqryylpsgdwsrlgryyvgiwqqeyzzsbibopnwaxwodejiekhbxflabtqwjfoyobnpettypktkvjgdzoghlbquhgjgufffmkeqvraawmuyymokbyhvjrbztshgdfylxfvvuddlvwchvsfopstqzgmmuhsbyswjtreftsarkxajrjujosbzkrhvzcebedtzxgirnsphvbgynkqfgmdgsfgkjpuoojvjsbrqodphfdbzsspsikbtasfuovbchvtrhpbmezqitfedribfiwhaendtywrejvedceqgddtimicnaidtzbcdsvggrvhihsbpgbeeqmkzpgi
```

Comment:
Also, what do you think of adding a few non-ASCII characters in here to (possibly?) improve the kind of data the benchmarks use?

Reply:
It doesn't matter which characters are included. It matters more what their distribution is. I chose a-z somewhat arbitrarily. I could easily have non-ascii characters, but the code treats the data as a `Vec<u8>` regardless.

Generally, noisy data is not a very good test case because almost all characters will be different despite common use cases including comparisons of strings that are very similar. I restricted the randomness to 26 characters because at least then there's a small chance for strings to have some similarities. Maybe it would be better to use less total characters or generate more realistic input patterns for the text.

Other ideas for random data:
* random a-z but biased by frequency in English prose
* random sequences of the top 10-20 English words

I'm interested in your ideas in this regard. My intuition is that these ideas won't affect the general trend of which function is faster, but might affect the run time in each case.

Reply:
Right, I see what you mean about about noisy data not being a good test case.

Considering common use cases of finding string distance, you'd generally expect that the things being compared are pretty similar, so I think the benchmarks should reflect that to some extent at least. So I'd suggest sequences of similar words (or words where the second word is a typo differing from the first word by a small amount). Generating this dataset shouldn't be hard at all. I'm open to merging this PR without this for now and adding better benchmarking data in a future PR too, cuz this is a good first step already.

That being said, the random noise you've generated is also pretty useful because you'd expect that the two sequences being compared are almost entirely different, so it turns into an exercise of the worst case inputs for the algorithm.

Reply:
> My intuition is that these ideas won't affect the general trend of which function is faster, but might affect the run time in each case.

This makes pretty good sense. However, I wouldn't be entirely surprised if the difference between how much faster tabulation is vs memoization varies significantly for some cases (e.g. when we use small sized strings, or when the sequences are mostly similar).

For example, the memoization implementation is likely this much slower because of the overhead of function calls, which tabulation doesn't have. This overhead might be quite insignificant if input sizes are really small, so the 3-5x difference between memoization and tabulation likely reduces by a large margin for these input sizes. Of course, I'm only speculating.

I think this further makes a case for having predefined similar sequences in our benchmarks. And also for using smaller input sizes (10, like we did for the naive recursive implementation, might work.

Thoughts?

Reply:
> This overhead might be quite insignificant if input sizes are really small, so the 3-5x difference between memoization and tabulation likely reduces by a large margin for these input sizes. Of course, I'm only speculating.

My tests show the factor is pretty consistent across input sizes.

> I think this further makes a case for having predefined similar sequences in our benchmarks. And also for using smaller input sizes (10, like we did for the naive recursive implementation, might work.

We do test input size 10 for tabulation/memoization already. I agree it would be useful to test other input data sets including ones that have more 'similar sequences'. However, it is not trivial to generate a variety of realistic data sets capturing different use cases. I'm leaning towards leaving that for future work down the line.

I respect the goal of this repo to demonstrate different techniques for calculating Levenshtein distance. However, all serious implementations of Levenshtein distance use a 'tabulation' algorithm because it is simply the fastest for real hardware. I think the random data case in these benchmarks is sufficient to show this. If additional data sets are added to test performance in the future, I don't think it makes sense to bother testing _all_ cases against _all_ three Levenshtein techniques for _all_ input sizes. You get a multiplicative explosion that's more effort to maintain than it's worth.

Reply:
That makes sense. I understand this PR lays the groundwork for benchmarking what this lib provides. The more realistic scenarios can be done later if needed.